tsl/src/continuous_aggs/refresh.c

```diff
@@ -290,15 +291,71 @@ log_refresh_window(int elevel, const ContinuousAgg *cagg, const InternalTimeRang
 		 DatumGetCString(OidFunctionCall1(outfuncid, end_ts)));
 }
 
+#define MAX_INVALIDATIONS_PER_REFRESH 10
```

Comment:
The semantics here are different from `max_interval_per_job`. The old `max_interval_per_job` setting is actually already similar to the user- or policy-specified refresh window; i.e., user controls the max amount materialized by specifying a specific refresh window size.

This max value is about how we optimize the materialization of the specified refresh window. Pre-2.0, everything from the earliest invalidation and forward was re-materialized no matter whether the entire range needed re-materialization or not. Post-2.0 we break that down so that we only materialize the regions/buckets that are actually invalid. However, if there are a lot of small regions that are invalidated (e.g., every other bucket in the worst case), it is cheaper to merge that into a large range. The session variable below controls the threshold.